pkg/globalstats/stats.go

```diff
@@ -49,45 +53,81 @@ func Incr(name string, tags ...stats.Tag) {
 	engine.Incr(name, tags...)
 }
 
+func Disable() {
+	mut.Lock()
+	defer mut.Unlock()
+
+	if flusherStop != nil {
+		close(flusherStop)
+	}
+}
+
 func Initialize(ctx context.Context, config Config) {
-	once.Do(func() {
-		if config.AppName == "" {
+	mut.Lock()
+	defer mut.Unlock()
+
+	if config.AppName == "" {
+		if len(os.Args) > 0 {
+			config.AppName = filepath.Base(os.Args[0])
+		} else {
 			config.AppName = "unknown"
 		}
-		if config.SamplePct <= 0 || config.SamplePct > 1 {
-			// by default only sample 10% of the observations
-			config.SamplePct = 0.10
+	}
+	if config.SamplePct == 0 {
+		// By default, only sample 10% of the observations.
+		config.SamplePct = 0.10
+	}
+	if config.CtlstoreVersion == "" {
+		config.CtlstoreVersion = ctlstore.Version
+	}
+	if config.FlushEvery == 0 {
+		config.FlushEvery = 10 * time.Second
+	}
+
+	samplePct = config.SamplePct
+
+	err := func() error {
+		if config.StatsHandler == nil {
+			return errors.New("no datadog client supplied")
+		}
+		if config.SamplePct > 1 || config.SamplePct < 0 {
+			return errors.New("sample percentage must be in the range of (0, 1]")
```

Comment:
**Note**: Just calling out this (small) breaking change that I'm happy to revert -- we now error if an invalid `SamplePct` or `FlushEvery` is given. IMO, this is important for `SamplePct` since someone could easily provide a percent as an integer instead of a decimal (5 instead of .05) which would lead to an unexpected frequency of stats.

Reply:
up the stack, we simply log errors from globalstats initialization and then go on our merry way. i wonder if we should terminate the program in some of these circumstances?

Reply:
Looks like all callsites go through [`ctlstore.Initialize`](https://sourcegraph.segment.com/search?q=ctlstore%5C.Initialize+-repo:segmentio/ctlstore+-repo:segmentio/ctlstore-legacy+-file:%28%5E%7C/%29vendor/+) ([I don't see anyone using `globalstats.Initialize` directly](https://sourcegraph.segment.com/search?q=globalstats%5C.Initialize+-repo:segmentio/ctlstore+-repo:segmentio/ctlstore-legacy+-file:%28%5E%7C/%29vendor/+-repo:segmentio/flagon)), so we should be good to make this panic instead. I'd be +1 on that

Reply:
We've apparently got a number of other programs using `globalstats.Initiailize`: https://sourcegraph.segment.com/search?q=globalstats.Initialize

Thought about this for a while and I think for now we can continue our current non-terminating behavior, out of an abundance of caution.  I think if it were a matter of configuring something more critical than instrumentation we could make a better argument for terminating.

Reply:
@collinvandyck see the link in my comment above, all of those use cases are either `ctlstore` or git-vendored versions of `ctlstore` or `flagon`

In that case, I believe we're fine to panic? But CMIIW :)

Reply:
Right but what happens when they update dependencies etc.  I think we should probably not panic in this case but log and return without setting up the engine et al.

Reply:
All of these callsites are by `ctlstore` in `ctlstore.Initialize`? Can you link me to an example of a non-`ctlstore` callsite of `globalstats.Initialize`?